src/ir/possible-contents.cpp

```diff
@@ -0,0 +1,1636 @@
+  }
+
+#ifdef POSSIBLE_CONTENTS_DEBUG
+  std::cout << "Link-targets phase\n";
+#endif
+
+  // Add all links to the targets vectors of the source locations, which we will
+  // use during the flow.
+  for (auto& link : links) {
+    getTargets(link.from).push_back(link.to);
+  }
+
+#ifndef NDEBUG
+  // Each vector of targets (which is a vector for efficiency) must have no
+  // duplicates.
+  for (auto& info : locations) {
+    disallowDuplicates(info.targets);
+  }
+#endif
+
+#ifdef POSSIBLE_CONTENTS_DEBUG
+  std::cout << "roots phase\n";
+#endif
+
+  // Set up the roots, which are the starting state for the flow analysis: send
+  // their initial content to them to start the flow.
+  for (const auto& [location, value] : roots) {
+#if defined(POSSIBLE_CONTENTS_DEBUG) && POSSIBLE_CONTENTS_DEBUG >= 2
+    std::cout << "  init root\n";
+    dump(location);
+    value.dump(std::cout, &wasm);
+    std::cout << '\n';
+#endif
+
+    updateContents(location, value);
+  }
+
+#ifdef POSSIBLE_CONTENTS_DEBUG
+  std::cout << "flow phase\n";
+  size_t iters = 0;
+#endif
+
+  // Flow the data while there is still stuff flowing.
+  while (!workQueue.empty()) {
+#ifdef POSSIBLE_CONTENTS_DEBUG
+    iters++;
+    if ((iters & 255) == 0) {
+      std::cout << iters++ << " iters, work left: " << workQueue.size() << '\n';
+    }
+#endif
+
+    auto iter = workQueue.begin();
+    auto locationIndex = *iter;
+    workQueue.erase(iter);
+
+    flowAfterUpdate(locationIndex);
+  }
+
+  // TODO: Add analysis and retrieval logic for fields of immutable globals,
+  //       including multiple levels of depth (necessary for itables in j2wasm).
+}
+
+bool Flower::updateContents(LocationIndex locationIndex,
+                            PossibleContents newContents) {
+  auto& contents = getContents(locationIndex);
+  auto oldContents = contents;
+
+#if defined(POSSIBLE_CONTENTS_DEBUG) && POSSIBLE_CONTENTS_DEBUG >= 2
+  std::cout << "updateContents\n";
+  dump(getLocation(locationIndex));
+  contents.dump(std::cout, &wasm);
+  std::cout << "\n with new contents \n";
+  newContents.dump(std::cout, &wasm);
+  std::cout << '\n';
+#endif
+
+  contents.combine(newContents);
+
+  if (contents.isNone()) {
+    // There is still nothing here. There is nothing more to do here but to
+    // return that it is worth sending more.
+    return true;
+  }
+
+  // It is not worth sending any more to this location if we are now in the
+  // worst possible case, as no future value could cause any change.
+  //
+  // Many is always the worst possible case. An exact type of a non-reference is
+  // also the worst case, since subtyping is not relevant there, and so if we
+  // know only the type then we already know nothing beyond what the type in the
+  // wasm tells us (and from there we can only go to Many).
+  bool worthSendingMore = !contents.isMany();
+  if (!contents.getType().isRef() && contents.isExactType()) {
+    worthSendingMore = false;
+  }
```

Comment:
> An exact type of a non-reference is also the worst case,

Why? Sorry I think I'm asking the same question over and over... Maybe we should take this and some other remaining questions offline, because I feel I'm holding this PR back for too long time.

The simple mental model I had in mind when I was reading this is a simple LUB kind of model, so you start from ⊥, which is `None`, and you combine types, and then it's possible end state is ⊤, which is `Many`.

Reply:
We talked about this offline and I summarized what I understood in https://github.com/WebAssembly/binaryen/pull/4685#discussion_r891979161.